CI: Run sanitizers on the entire test corpus

reneme opened the pull request:
This adds CI jobs running clang's ASan and UBsan as well as MSVC's ASan and checked iterators on the entire test corpus (most notably BoGo tests). This revealed a memory leak in the `bogo_shim`.

Note that parallel building (via `jom`) with debug symbols did not work on windows. This would require multiple processes to access `botan.pdb` concurrently which fails. The suggested fix (adding `/FS`) didn't work either -- likely due to lacking support of file-system locking in GitHub Action's VMs. Therefore, the windows sanitizer build uses `nmake` (and therefore builds sequentially).

This also uses `BOTAN_FFI_RETURNING` more consistently in `ffi_cert.cpp` to silence warnings showing up in the MSVC sanitizer build. See [here for details on this macro](https://github.com/randombit/botan/blob/master/src/lib/ffi/ffi_util.h#L96-L107).

lieser commented on src/scripts/ci_build.py:
```diff
@@ -152,7 +154,12 @@ def determine_flags(target, target_os, target_cpu, target_cc, cc_bin,
+        flags += ['--with-coverage-info']
+
+    if target in ['coverage', 'sanitizer']:
+        if target_os != 'windows':
+            flags += ['--with-debug-info']
```
> Shall we try it?


I would say yes

reneme replied:
[Seems to work just fine](https://github.com/randombit/botan/actions/runs/3548877661/jobs/5960581074) without this. Some macOS-based configurations failed as there seems to be some outage at homebrew.

lieser replied:
You commented out the code, but I think it should be removed

lieser replied:
Also you now removed `--with-debug-info` now completely. But for nice outputs we probably want debug info in our sanitizer builds.
So I think it should be added for all, not removed for all.

reneme replied:
[Another bread-crumb](https://stackoverflow.com/a/21855956): Someone had this error while building inside a Dropbox share, because the dropbox daemon locked files at will and interfered with the build. Maybe some anti-virus software is meddling with the file.

lieser commented:
For me it is currently a little unclear what the distinguishing between the `fuzzers` and `sanitizer` job is supposed to be.

After this PR we would have:

1. fuzzers, gcc
2. sanitizer, clang
3. sanitizer, msvc

Some observations/questions:

- All three are building the test fuzzer. But only the fuzzers job is actually executing them.
- The fuzzers job is also executing at least part of the normal tests, but not e.g. the bogo tests. Is this intended and needed?
- The sanitizer jobs are also running the bogo and cli tests
- The fuzzer job is also using sanitizers. Is the sanitizer config different from the sanitizer job? Or why do we need/want to distinguish between them?

Seems to be currently like we probably still want to have three such jobs (one for each compiler), but let them run the same tests and name them the same (I would tend to name them sanitizer).

randombit commented:
The intent of the fuzzers build is just to build and run the fuzzers to ensure they work. The reason for using the sanitizers there is that this is typically how the sanitizers are built (eg in OSS-Fuzz) and will detect any memory errors caused by the corpus itself.

>  All three are building the test fuzzer.

I don't think that's correct. If it is correct, it was certainly unintentional!

```
    if target in ['fuzzers', 'coverage']:
        flags += ['--build-fuzzers=test']
```

> The fuzzers job is also executing at least part of the normal tests, but not e.g. the bogo tests. Is this intended and needed?

It's certainly intentional that we skip the BoGo tests in the fuzzer job since that doesn't relate to the fuzzers. It actually would be better if the fuzzers job skipped the normal tests as well, is it is redundant with the other builds.

> Seems to be currently like we probably still want to have three such jobs (one for each compiler), but let them run the same tests and name them the same (I would tend to name them sanitizer).

It would be fine to have an equivalent to the Clang sanitizer build for GCC. That might be helpful as occasionally GCC's ASan behaves a little differently from Clang's. ISTR a couple of bugs that were caught by one but not the other. But there is no need to build and run the fuzzers under all three compilers IMO.

Also the fuzzers job enables `--unsafe-fuzzer-mode` which is not in general how we'd want to test the library, since it disables various signature verifications and so on.

lieser commented on src/scripts/ci_build.py:
```diff
@@ -178,7 +185,7 @@ def determine_flags(target, target_os, target_cpu, target_cc, cc_bin,
         flags += ['--unsafe-fuzzer-mode', '--terminate-on-asserts']
 
-    if target in ['fuzzers', 'coverage']:
+    if target in ['fuzzers', 'coverage', 'sanitizer']:
         flags += ['--build-fuzzers=test']
```
With the clarifications of the differences between the fuzzers and sanitizer jobs is does not seem like we want to build the fuzzer tests also for the sanitizer jobs.

reneme replied:
That makes sense to me, too. Changed.

reneme commented:
I removed the `--build-fuzzers=test` flag from `sanitizer` builds and added a GCC sanitizer job. Maybe it would be helpful when the CI pipeline would generate some sort of input/output matrix and publish that as a build result. I'd hope that this would provide a quick overview over which job is building and running what. Something along the lines of:

| CI Flag | sanitizer/gcc | sanitizer/clang | fuzzers/gcc | coverage | ... |
| --- | --- | --- | --- | --- | --- |
| `--build-fuzzers=test` | | | :heavy_check_mark: | :heavy_check_mark: | |
| `--enable-sanitizers=address,undefined` | :heavy_check_mark: | :heavy_check_mark: | | | |
| `--with-debug-asserts` | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | | |
| ... | | | | | |

lieser reviewed (changes requested):
I think the `--with-debug-info` discussion is still open.

reneme commented:
Note: This cannot be merged, yet. MSVC complained about more locations where `BOTAN_FFI_DO` should be replaced by `BOTAN_FFI_RETURNING`. Though, now that we have C++17 we can do better than this workaround and let the compiler figure out what to do: https://github.com/randombit/botan/pull/3079

Let's merge #3079 first and rebase this to get the CI to pass.

reneme commented:
Rebased after #3079 was merged. Please be nice, MSVC!

reneme commented on src/scripts/ci_build.py:
```diff
@@ -201,7 +206,7 @@ def sanitize_kv(some_string):
             flags += ['--enable-sanitizers=address,undefined']
         else:
-            flags += ['--with-sanitizers']
+            flags += ['--enable-sanitizers=address']
```
This explicitly disables the `-D_ITERATOR_DEBUG_LEVEL=1` as it doesn't seem to be compatible with ASan. Linking of the botan library fails with:

```
VCASAN.lib(vcasan.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '1' in asn1_alg_id.obj
   Creating library .\botan.lib and object .\botan.exp
.\botan.dll : fatal error LNK1319: 1 mismatches detected
```

randombit commented on .github/workflows/ci.yml:
```diff
@@ -155,7 +155,16 @@ jobs:
+                             # See: https://github.com/randombit/botan/pull/3012#discussion_r1033621569
+          - target: sanitizer
+            compiler: clang
+            # in Oct. 2022 'ubuntu-latest' would mean 'ubuntu-20.04' only
```
These comments can be removed since we are on `ubuntu-22.04` for all CI now

randombit reviewed (approved):
Changes look good but there is still a merge conflict and some CI failures. Also please rebase+squash the commits before merge.

reneme commented:
Rebased, squased and now waiting for CI. I will force push again tomorrow. Don't have my YubiKey for commit signing handy. >.<

randombit commented on src/scripts/ci_build.py:
```diff
@@ -326,6 +331,10 @@ def sanitize_kv(some_string):
+        if target_os != 'windows' or target != 'sanitizer':
+            # GitHub Actions worker intermittently ran out of memory when
+            # asked to allocate multi-gigabyte buffers under MSVC's ASan.
+            test_cmd += ['--run-memory-intensive-tests']
```
Let's invert this option (to `--skip-memory-intensive-tests`) because otherwise nobody (including eg an end user building on some weird CPU/compiler we've never seen) will have these tests run when they build locally, unless they know about this magic flag. That doesn't matter so much for this one particular RNG test but would be if it started being used to gate for instance an Argon2 test with very large memory.

reneme replied:
I considered that for the same reason but went with the 'disabled by default' to stay consistent with `--run-long-tests`. We could turn both around? Or just don't care about consistency here. I don't have a strong feeling either way.

randombit replied:
Eh. Not really strong feelings here I guess.

Let's make one minor change here: add `--run-memory-intensive-tests` to the default CI run and then *remove* it for the Windows sanitizer build only. That way we are still running the RNG 4GB test on macOS and Windows. Otherwise, AIUI, we are only ever running it within the Linux coverage build, so we could miss some problem that creeps into how we implement `System_RNG` on another platform.

reneme replied:
Done. For the record: The same is true for the "long" tests. `--run-long-tests` is only ever executed for the coverage build (and -- with this PR -- with sanitizers on all three platforms)

randombit replied:
Thanks for pointing that out. I'll address this.

randombit reviewed (approved):
Approved modulo the option comment
